lib/charms/ip_router_interface/v0/ip_router_interface.py

```diff
@@ -29,4 +132,273 @@
+            self._stored.routing_table.pop(event.app.name)
+            self._sync_routing_tables()
+
+    def get_routing_table(self) -> RoutingTable:
+        """Read-only way to get the current routing table"""
+        return deepcopy(self._stored.routing_table._under)  # RFC: Is there a better way?
+
+    def get_flattened_routing_table(self) -> List[Network]:
+        """Returns a read-only internal routing table that's flattened
+
+        Returns:
+            A list of objects of type `Network`
+        """
+        internal_rt = self.get_routing_table()
+        final_rt = []
+        for networks in internal_rt.values():
+            if type(networks) is not str:
+                continue
+            for network in json.loads(networks):
+                final_rt.append(network)
+
+        return final_rt
+
+    def _sync_routing_tables(self) -> None:
+        """Syncs the internal routing table with all of the requirer's app databags"""
+        ip_router_relations = self.model.relations["ip-router"]
+        for relation in ip_router_relations:
+            relation.data[self.charm.app].update(
+                {"networks": json.dumps(self.get_flattened_routing_table())}
+            )
+
+
+class RouterRequires(Object):
+    """ip-router requirer class to be instantiated by charms that require routing
+
+    This class provides methods to request a new network, and read the available
+    network from the router providers. These should be used exclusively to
+    interact with the relation.
+
+    Attributes:
+        charm: The Charm object that instantiates this class.
+    """
+
+    def __init__(self, charm: CharmBase):
+        super().__init__(charm, "ip-router")
+        self.charm = charm
+
+    def request_network(
+        self,
+        networks: List[Network],
+    ) -> None:
+        """Requests a new network interface from all of the ip-router providers
+
+        The interfaces must be valid according to `_network_is_valid`. Multiple
+        calls to this function will replace the previously requested networks,
+        so all of the networks required must be given with each call.
+
+        Arguments:
+            networks:
+                A list containing the desired networks of the type `Network`.
+
+        Raises:
+            RuntimeError:
+                No ip-router relation exists yet or validation of one
+            or more of the networks failed.
+        """
+        if not self.charm.unit.is_leader():
+            return
+
+        ip_router_relations = self.model.relations.get("ip-router")
+        if len(ip_router_relations) == 0:
+            raise RuntimeError("No ip-router relation exists yet.")
+
+        for network_request in networks:
+            try:
+                self._validate_network(network_request, networks)
+            except (KeyError, ValueError) as e:
+                raise RuntimeError(f"{network_request} failed to validate.", str(e))
+
+        # Place it in the databags
```

Comment:
Not necessary

Reply:
Removed

Reply:
It's updated to do logging instead of raising another error, if you click the link to the file from above this comment, it will take you to the old version of the PR. You must select the `changes from all commits` manually.